### PR TITLE
Adding a global token URI to Unlock

### PR DESCRIPTION
**Pull request by nfurfaro**
# Description

This is the first of a few PR's related to implementing metadata for Unlock. 
This one focuses on `Unlock.sol` and adds a storage field and getter/setter funcs.

# Issues

Fixes #
Refs #1764

# Checklist:

- [x] 1 PR, 1 purpose: my Pull Request applies to a single purpose
  - [ ] This PR only contains configuration changes (package.json, etc.)
  - [x] This PR only contains code changes (if configuration changes are required, do a separate PR first, then re-base)
- [x] My code follows the style guidelines of this project, including naming conventions
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] If my code adds or changes components, I have written corresponding stories with Storybook
- [x] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

File: smart-contracts/contracts/Unlock.sol [hunk 1]
```diff
@@ -80,6 +84,7 @@ contract Unlock is
   {
     // We must manually initialize Ownable.sol
     Ownable.initialize(_owner);
+    globalBaseTokenURI = 'https://locksmith.unlock-protocol.com/api/key/';
```
Comment (hardlydifficult) [inline smart-contracts/contracts/Unlock.sol]: We can't initialize here, it won't be called again.  How about no default in the contract, and then we use the changeBase function to set it?

File: smart-contracts/contracts/Unlock.sol [hunk 2]
```diff
@@ -183,4 +188,24 @@ contract Unlock is
...
+
+
+  // function to set the globalTokenURI field.
+  function setGlobalBaseTokenURI(
+    string calldata _URI
+  )
+    external
+    onlyOwner
+  {
+    globalBaseTokenURI = _URI;
```
Comment (hardlydifficult) [inline smart-contracts/contracts/Unlock.sol]: Should we emit an event here?  I'm actually not sure if it's worth it... but I believe we do emit events for all other state changes.

File: smart-contracts/contracts/Unlock.sol (hunk 1 above)
Comment (nfurfaro) [reply smart-contracts/contracts/Unlock.sol]: of course! ha, this only works in testing. yep, will remove this and we'll have to set it explicitly with the setter func.

File: smart-contracts/contracts/Unlock.sol (hunk 2 above)
Comment (nfurfaro) [reply smart-contracts/contracts/Unlock.sol]: Yeah, I think it's good practice in general. I'll add & emit an event.

**Comment by nfurfaro** [comment]
OK, this is weird...Was just trying to add an event and ran into this.
right now I have 3 tests that pass (in the new test file), and no events.
```
  Contract: Lock / erc721 / tokenURI
    the global tokenURI stored in Unlock
      ✓ should return the global base token URI
      ✓ should allow the owner to set the global token URI (63ms)
      ✓ should fail if someone other than the owner tries to set the URI (48ms)


  3 passing (238ms)
```
But, as soon as I add this below `event NewLock`:
```
event NewTokenURI(
    string tokenURI
  );
```
This is the result of running the same test:
```
 Contract: Lock / erc721 / tokenURI
    the global tokenURI stored in Unlock
      1) should return the global base token URI
    > No events were emitted
      2) should allow the owner to set the global token URI
    > No events were emitted
      ✓ should fail if someone other than the owner tries to set the URI (83ms)


  1 passing (2s)
  2 failing

  1) Contract: Lock / erc721 / tokenURI
       the global tokenURI stored in Unlock
         should return the global base token URI:
     Error: Invalid number of parameters for "getGlobalBaseTokenURI". Got 0 expected 1!
      at Object.InvalidNumberOfParams (node_modules/truffle/build/webpack:/~/web3-eth-contract/~/web3-core-helpers/src/errors.js:32:1)
      at Object._createTxObject (node_modules/truffle/build/webpack:/~/web3-eth-contract/src/index.js:699:1)
      at Promise (node_modules/truffle/build/webpack:/packages/truffle-contract/lib/execute.js:128:1)
      at process.internalTickCallback (internal/process/next_tick.js:77:7)

  2) Contract: Lock / erc721 / tokenURI
       the global tokenURI stored in Unlock
         should allow the owner to set the global token URI:
     Error: Invalid number of parameters for "getGlobalBaseTokenURI". Got 0 expected 1!
      at Object.InvalidNumberOfParams (node_modules/truffle/build/webpack:/~/web3-eth-contract/~/web3-core-helpers/src/errors.js:32:1)
      at Object._createTxObject (node_modules/truffle/build/webpack:/~/web3-eth-contract/src/index.js:699:1)
      at Promise (node_modules/truffle/build/webpack:/packages/truffle-contract/lib/execute.js:128:1)
      at process.internalTickCallback (internal/process/next_tick.js:77:7)
```
No other changes were made. Commenting out the new event gets all tersts passing again.
Thoughts? @hardlydifficult

File: smart-contracts/contracts/Unlock.sol (hunk 1 above)
Comment (nfurfaro) [reply smart-contracts/contracts/Unlock.sol]: done

**Comment by hardlydifficult** [comment]
hmm no repro? (see last commit)

is this because of the build and/or ganache being out of date maybe?

**Comment by nfurfaro** [comment]
Last commit doesn't have the new event yet. Will try nuking and installing again.

**Comment by nfurfaro** [comment]
Ok, think this is good to go.

File: smart-contracts/contracts/Unlock.sol [hunk 3]
```diff
@@ -183,4 +191,25 @@ contract Unlock is
   {
     return 2;
   }
-}
+
+  // function to read the globalTokenURI field.
+  function getGlobalBaseTokenURI()
+    public
```
Comment (hardlydifficult) [inline smart-contracts/contracts/Unlock.sol]: external?

Comment (nfurfaro) [reply smart-contracts/contracts/Unlock.sol]: fixed.

**Pull request merged**